backend/src/main/java/org/polsl/backend/service/SoftwareService.java

```diff
@@ -33,4 +52,52 @@ public SoftwareService(SoftwareRepository softwareRepository) {
     response.setTotalElements((long) softwareOutputDTO.size());
     return response;
   }
+
+  @Transactional
+  public void createSoftware(SoftwareInputDTO request) {
+    Software software = new Software();
+    software.setName(request.getName());
+    softwareRepository.save(software);
+
+    Set<Long> computerSetIdsSet = request.getComputerSetIds();
+    if (computerSetIdsSet != null) {
+      computerSetIdsSet.forEach(computerSetId -> {
+        ComputerSet computerSet = computerSetRepository.findById(computerSetId)
+            .orElseThrow(() -> new NotFoundException("zestaw komputerowy", "id", computerSetId));
+        ComputerSetSoftware computerSetSoftware = new ComputerSetSoftware(computerSet, software);
+        computerSetSoftwareRepository.save(computerSetSoftware);
+      });
+    }
+  }
+
+  @Transactional
+  public void editSoftware(Long id, SoftwareInputDTO request) throws NotFoundException {
+    //Edit name
+    Software software = softwareRepository.findById(id).orElseThrow(() -> new NotFoundException("oprogramowanie", "id", id));
+    software.setName(request.getName());
+    softwareRepository.save(software);
+
+    Set<Long> computerSetIdsSet = request.getComputerSetIds();
+    if (computerSetIdsSet != null) {
+      Set<ComputerSetSoftware> computerSetSoftwareSet = computerSetSoftwareRepository.findAllBySoftwareId(id);
+
+      //Old record(s)
+      for (ComputerSetSoftware computerSetSoftware : computerSetSoftwareSet) {
+        computerSetSoftware.setValidTo(LocalDateTime.now());
+        computerSetSoftwareRepository.save(computerSetSoftware);
+      }
```

Comment:
W tym momencie edytujesz absolutnie wszystkie wpisy historii dotyczące połączenia oprogramowania z zestawami. Czyli również jakieś wpisy z odległej przeszłości, które mają już ustawioną konkretną `valid_to`. A my musimy zmienić tutaj tylko ostatnie wpisy, czyli dla każdej pary (`computerSetId`, `softwareId`) trzeba wyszukać przez repozytorium tylko najnowszy wpis. Jeśli by nie dało się tego zrobić przez zwykłą metodę CrudRepository, to by trzeba było się wspomóc adnotacją `@Query`.

Dobrze by było też wykonywać taką operację tylko wtedy, kiedy ID zestawu komputerowego się zmienia. To znaczy: jeśli wcześniej oprogramowanie było przypisane do zestawów o id = [1, 2], a zmienia się na [2, 3], to w rezultacie:
- `valid_to` rekordu o id = 1 powinien zostać ustawiony na bieżącą datę
- rekord w tabeli dotyczący zestawu o id = 2 powinien zostać nietknięty
- powinien zostać utworzony nowy rekord dotyczący id = 3

Mega skomplikowane 😟, ale jeśli teraz zrobimy to w 100% dobrze, to potem nam się to opłaci 😎

Reply:
Poprawione